src/Storages/Streaming/StreamingStoreSourceBase.h

```diff
@@ -50,7 +50,8 @@ class StreamingStoreSourceBase : public ISource
 
     Int64 last_sn = -1;
     Int64 last_epoch = -1;
-    /// FIXME, switch to llvm-15
-    std::atomic<CheckpointContext *> ckpt_ctx;
+    /// FIXME, switch to llvm-15 atomic shared_ptr
+    CheckpointContextPtr ckpt_ctx TSA_PT_GUARDED_BY(ckpt_mutex);
+    std::mutex ckpt_mutex;
```

Comment:
We may need apply 
1) fast mutex 
2) double checked locking to avoid lock a mutex always.

```
std::atomic<bool> has_ckpt_request = false;
FastMutext ckpt_mutex;
```

```
/// Check
{
   CheckpointContextPtr ckpt;
    if (has_ckpt_request)
    {
        std::scope_lock lock(ckpt_mutex);
        if (ckpt_ctx)
        {
            ckpt = std::move(ckpt_ctx); /// We moved it away, so no need to call ckpt_ctx.reset()
            has_ckpt_request = false;
        }
    }
}


/// Set
{
    /// ckpt coordinator guarantees that ckpt is not overlapped.
    assert (!has_ckpt_request)
    std::scope_lock lock(ckpt_mutex);
    ckpt_ctx = std::make_shared<CheckpointContext>(*ckpt_ctx_);
    has_ckpt_request = true;
}
```

Reply:
this is just a demo about mutex and atomicbool check.not ready merge.

micro benchmark I will
do in later . 

I intend to use atomic<ckpt*> and std mutex.